modules/index/src/main/java/org/apache/ignite/internal/index/IndexAvailabilityController.java

```diff
@@ -0,0 +1,327 @@
+                    ),
+                    List.of(noop())
+            );
+        });
+    }
+
+    private CompletableFuture<?> onIndexDrop(DropIndexEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            int indexId = parameters.indexId();
+
+            int partitions = getPartitionCountFromCatalog(indexId, parameters.catalogVersion() - 1);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            return metaStorageManager.invoke(
+                    exists(startBuildIndexKey),
+                    concat(
+                            List.of(remove(startBuildIndexKey)),
+                            removePartitionBuildIndexOperations(indexId, partitions)
+                    ),
+                    List.of(noop())
+            );
+        });
+    }
+
+    private CompletableFuture<?> onIndexAvailable(MakeIndexAvailableEventParameters parameters) {
+        return inBusyLockAsync(busyLock, () -> {
+            ByteArray startBuildIndexKey = startBuildIndexKey(parameters.indexId());
+
+            return metaStorageManager.invoke(exists(startBuildIndexKey), remove(startBuildIndexKey), noop());
+        });
+    }
+
+    private CompletableFuture<?> onUpdatePartitionBuildIndexKey(WatchEvent event) {
+        return inBusyLockAsync(busyLock, () -> {
+            if (!event.single()) {
+                // We don't need to handle keys on index creation or deletion.
+                return completedFuture(null);
+            }
+
+            Entry entry = event.entryEvent().newEntry();
+
+            if (!entry.tombstone()) {
+                // In case an index was created when there was only one partition.
+                return completedFuture(null);
+            }
+
+            String partitionBuildIndexKey = new String(entry.key(), UTF_8);
+
+            int indexId = parseIndexIdFromPartitionBuildIndexKey(partitionBuildIndexKey);
+
+            ByteArray startBuildIndexKey = startBuildIndexKey(indexId);
+
+            long metastoreRevision = entry.revision();
+
+            if (isRemainingPartitionBuildIndexKeys(indexId, metastoreRevision)
+                    || isMetastoreKeyAbsent(startBuildIndexKey, metastoreRevision)) {
+                return completedFuture(null);
+            }
+
+            // We can use the latest version of the catalog since we are on the metastore thread.
+            CatalogIndexDescriptor indexDescriptor = getIndexDescriptorStrict(indexId, catalogManager.latestCatalogVersion());
+
+            // We will not wait for the command to be executed, since we will then find ourselves in a dead lock since we will not be able
+            // to free the metastore thread.
+            catalogManager.execute(buildMakeIndexAvailableCommand(indexDescriptor));
+
+            return completedFuture(null);
+        });
+    }
+
+    private void onIndexBuildForPartition(int indexId, int partitionId) {
```

Comment:
It seems that we now use **index build** and not **index built** everywhere, I’ll leave it as it is.

Reply:
I'm not an expert in English, but 'On index build' seems to mean that the index has started being built, or that it's being built right now. But this method is about the 'finish' event. So the method has to either have 'finish' (or its analogue) in its name, or it should mention the 'build' in the past ('built'), otherwise it confuses a reader.